File: 2223_SpinUp/Sparta/src/logistic.cpp
#include "logistic.h"
#include "okapi/api/units/QAngle.hpp"

Logistic::Logistic(double startTime, double numberOfIntervals, double thetaDifference, okapi::QAngle startTheta) {
    TIME_INCREMENT = abs(startTime) * 2 / numberOfIntervals;
    MAX_VAL = thetaDifference; // asymptotic to target theta, fastest at theta / 2

    currTime = startTime;
    idealTheta = 0_deg;
    this->startTheta = startTheta;
}

double Logistic::val(double max, double t) {
    return max / (1 + pow(M_E, -t));
}

double Logistic::derivative(double max, double t) {
    return max * (1 - val(1, t)) * val(1, t);
}

okapi::QAngle Logistic::step() {
    currTime += TIME_INCREMENT;

    idealTheta = val(MAX_VAL, currTime) * 1_deg;

    std::cout << idealTheta.convert(okapi::degree) << "\n";

    return startTheta + idealTheta;


    /*
    double targetTheta = 5;

    const double START_TIME = -5;
    const double TIME_INCREMENT = 0.1;

    double currMotorVel = val(targetTheta, START_TIME);

    for (double i = START_TIME; i <= -START_TIME; i += TIME_INCREMENT) {
        // std::cout << "(" << i << ", " << currMotorVel * TIME_INCREMENT << ")" << "\n";
        
        if (i < 0)
            currMotorVel += derivative(targetTheta, i - (TIME_INCREMENT / 2));
        else if (i >= 0)
            currMotorVel -= derivative(targetTheta, i - (TIME_INCREMENT / 2));
    }
    */
}

File: 2223_SpinUp/Sparta/LogisticTuning/logistic.py
from math import e


class Logistic:

    def __init__(self, _start_time, _num_intervals, _theta_diff, _start_theta):
        self.time_increment = abs(_start_time) * 2 / _num_intervals
        self.max_val = _theta_diff
        self.cur_time = _start_time
        self.ideal_theta = 0
        self.start_theta = _start_theta

    @staticmethod
    def logistic_val(max_output, time):
        return max_output / (1 + pow(e, -time))

    def step(self):
        self.cur_time += self.time_increment
        self.ideal_theta = Logistic.logistic_val(self.max_val, self.cur_time)
        return self.start_theta + self.ideal_theta
